flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/PreBatchTransformOperator.java

```diff
@@ -0,0 +1,389 @@
+        this.udfDescriptors =
+                this.udfFunctions.stream()
+                        .map(udf -> new UserDefinedFunctionDescriptor(udf.f0, udf.f1, udf.f2))
+                        .collect(Collectors.toList());
+
+        // Initialize data fields in advance because they might be accessed in
+        // `::initializeState` function when restoring from a previous state.
+        this.transforms = new ArrayList<>();
+        for (TransformRule transformRule : transformRules) {
+            String tableInclusions = transformRule.getTableInclusions();
+            String projection = transformRule.getProjection();
+            String filter = transformRule.getFilter();
+            String primaryKeys = transformRule.getPrimaryKey();
+            String partitionKeys = transformRule.getPartitionKey();
+            String tableOptions = transformRule.getTableOption();
+            Selectors selectors =
+                    new Selectors.SelectorsBuilder().includeTables(tableInclusions).build();
+            transforms.add(
+                    new PreTransformer(
+                            selectors,
+                            TransformProjection.of(projection).orElse(null),
+                            TransformFilter.of(filter, udfDescriptors).orElse(null)));
+            schemaMetadataTransformers.add(
+                    new Tuple2<>(
+                            selectors,
+                            new SchemaMetadataTransform(primaryKeys, partitionKeys, tableOptions)));
+        }
+        this.preTransformProcessorMap = new ConcurrentHashMap<>();
+        this.hasAsteriskMap = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public void finish() throws Exception {
+        super.finish();
+        clearOperator();
+    }
+
+    @Override
+    public void close() throws Exception {
+        super.close();
+        clearOperator();
+    }
+
+    @Override
+    public void processElement(StreamRecord<Event> element) throws Exception {
+        Event event = element.getValue();
+        if (event instanceof CreateTableEvent) {
+            CreateTableEvent createTableEvent = (CreateTableEvent) event;
+            preTransformProcessorMap.remove(createTableEvent.tableId());
+            output.collect(new StreamRecord<>(cacheCreateTable(createTableEvent)));
+        } else if (event instanceof DropTableEvent) {
+            preTransformProcessorMap.remove(((DropTableEvent) event).tableId());
+            output.collect(new StreamRecord<>(event));
+        } else if (event instanceof TruncateTableEvent) {
+            output.collect(new StreamRecord<>(event));
+        } else if (event instanceof SchemaChangeEvent) {
+            SchemaChangeEvent schemaChangeEvent = (SchemaChangeEvent) event;
+            preTransformProcessorMap.remove(schemaChangeEvent.tableId());
+            cacheChangeSchema(schemaChangeEvent)
+                    .ifPresent(e -> output.collect(new StreamRecord<>(e)));
```

Comment:
No DropTableEvent, TruncateTableEvent and other SchemaChangeEvents will be sent, so we can only deal with CreateTableEvent and DataChangeEvent.